go-controller/pkg/ovn/multipolicy_test.go

```diff
@@ -0,0 +1,543 @@
+			allportInfo: map[string]portInfo{},
+		}
+		p.secondaryPodInfos[netName] = podInfo
+	}
+	portName := util.GetSecondaryNetworkLogicalPortName(p.namespace, p.podName, nadName)
+	podInfo.allportInfo[nadName] = portInfo{
+		portUUID: portName + "-UUID",
+		podIP:    podIP,
+		podMAC:   podMAC,
+		portName: portName,
+	}
+}
+
+func (p testPod) getNetworkPortInfo(netName, nadName string) *portInfo {
+	podInfo, ok := p.secondaryPodInfos[netName]
+	if !ok {
+		return nil
+	}
+	info, ok := podInfo.allportInfo[nadName]
+	if !ok {
+		return nil
+	}
+
+	return &info
+}
+
+func addPodNetwork(pod *v1.Pod, secondaryPodInfos map[string]*secondaryPodInfo) {
+	nadNames := []string{}
+	for _, podInfo := range secondaryPodInfos {
+		for nadName := range podInfo.allportInfo {
+			nadNames = append(nadNames, nadName)
+		}
+	}
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+	pod.Annotations[nettypes.NetworkAttachmentAnnot] = strings.Join(nadNames, ",")
+}
+
+func (p testPod) populateSecondaryNetworkLogicalSwitchCache(fakeOvn *FakeOVN, ocInfo secondaryControllerInfo) {
+	var err error
+	switch ocInfo.topology {
+	case ovntypes.Layer3Topology:
+		l3Controller := ocInfo.NetworkController.(*SecondaryLayer3NetworkController)
+		podInfo := p.secondaryPodInfos[ocInfo.GetNetworkName()]
+		uuid := getLogicalSwitchUUID(fakeOvn.controller.nbClient, l3Controller.GetNetworkScopedName(p.nodeName))
+		err = l3Controller.lsManager.AddSwitch(l3Controller.GetNetworkScopedName(p.nodeName), uuid, []*net.IPNet{ovntest.MustParseIPNet(podInfo.nodeSubnet)})
+	case ovntypes.Layer2Topology:
+		l2Controller := ocInfo.NetworkController.(*SecondaryLayer2NetworkController)
+		uuid := getLogicalSwitchUUID(fakeOvn.controller.nbClient, l2Controller.GetNetworkScopedName(ovntypes.OVNLayer2Switch))
+		subnet := l2Controller.NetConfInfo.Subnets()[0]
+		err = l2Controller.lsManager.AddSwitch(l2Controller.GetNetworkScopedName(ovntypes.OVNLayer2Switch), uuid, []*net.IPNet{ovntest.MustParseIPNet(subnet)})
+	case ovntypes.LocalnetTopology:
+		localnetController := ocInfo.NetworkController.(*SecondaryLocalnetNetworkController)
+		uuid := getLogicalSwitchUUID(fakeOvn.controller.nbClient, localnetController.GetNetworkScopedName(ovntypes.OVNLayer2Switch))
+		subnet := localnetController.NetConfInfo.Subnets()[0]
+		err = localnetController.lsManager.AddSwitch(localnetController.GetNetworkScopedName(ovntypes.OVNLocalnetSwitch), uuid, []*net.IPNet{ovntest.MustParseIPNet(subnet)})
+	}
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+}
+
+func getExpectedDataPodsAndSwitchesForSecondaryNetwork(fakeOvn *FakeOVN, pods []testPod) []libovsdb.TestData {
+	data := []libovsdb.TestData{}
+	for _, ocInfo := range fakeOvn.secondaryControllers {
+		nodeslsps := make(map[string][]string)
+		var bnc *BaseNetworkController
+		var switchName string
+		switch ocInfo.topology {
+		case ovntypes.Layer3Topology:
+			l3Controller := ocInfo.NetworkController.(*SecondaryLayer3NetworkController)
```

Comment:
we might as well embed or make NetInfo retrievable from the NetwrokController interface

Reply:
how can that help. We'd still need its embeded BaseSecondaryNetworkController to watch for namepsace, pod and multi-network-policy.

Reply:
I was just looking at this method. 

But indeed, if there is multiple reasons to do this, then I think it is better to encapsulate these type of things in `FakeOVN` itself. Otherwise every other test suite will need to replicate this code.

For example, this `getExpectedDataPodsAndSwitchesForSecondaryNetwork` looks like something that any test suite that deals with secondary networks will need.

Also, FakeOVN could just offer a method `getNetworkController(networkName)` that returns a test specific interface with the methods needed like `TopologyType` or `WatchNamespaces`. If not all the network controllers implement all methods it could just be an adapter that does nothing for the methods not implemented.

Reply:
I've changed the FakeOVN.secondaryControllers to be map of secondary network controller information, which includes the BaseSecondaryNetworkController. It should be easily used by any other consumer if needed.